apps/api-extractor/src/generators/PackageTypingsGenerator.ts

```diff
@@ -185,65 +204,108 @@ export default class PackageTypingsGenerator {
     while (true) { // tslint:disable-line:no-constant-condition
 
       for (const declaration of current.declarations || []) {
+        // 1. Check for any signs that this is not an ambient definition
         if (declaration.kind === ts.SyntaxKind.ExportSpecifier
           || declaration.kind === ts.SyntaxKind.ExportAssignment) {
           isAmbient = false;
-          break;
         }
 
         const modifiers: ts.ModifierFlags = ts.getCombinedModifierFlags(declaration);
         if (modifiers & (ts.ModifierFlags.Export | ts.ModifierFlags.ExportDefault)) {
-            isAmbient = false;
-            break;
+          isAmbient = false;
+        }
+
+        // 2. Check for any signs that this was imported from an external package
+        {
+          // EXAMPLE:
+          // "export { A } from './file-a';"
+          //
+          // ExportDeclaration:
+          //   ExportKeyword:  pre=[export] sep=[ ]
+          //   NamedExports:
+          //     FirstPunctuation:  pre=[{] sep=[ ]
+          //     SyntaxList:
+          //       ExportSpecifier:  <------------- declaration
+          //         Identifier:  pre=[A] sep=[ ]
+          //     CloseBraceToken:  pre=[}] sep=[ ]
+          //   FromKeyword:  pre=[from] sep=[ ]
+          //   StringLiteral:  pre=['./file-a']
+          //   SemicolonToken:  pre=[;]
+          const exportDeclaration: ts.ExportDeclaration | undefined
+            = PackageTypingsGenerator._matchAncestor<ts.ExportDeclaration>(declaration,
+            [ts.SyntaxKind.ExportDeclaration, ts.SyntaxKind.NamedExports, ts.SyntaxKind.ExportSpecifier]);
+
+          if (exportDeclaration && exportDeclaration.moduleSpecifier) {
+            // Examples:
+            //    " '@microsoft/sp-lodash-subset'"
+            //    " "lodash/has""
+            const packagePath: string | undefined = PackageTypingsGenerator._getPackagePathFromModuleSpecifier(
+              exportDeclaration.moduleSpecifier);
+
+            if (packagePath) {
+              // Example: " ExportName as RenamedName"
+              const exportSpecifier: ts.ExportSpecifier = declaration as ts.ExportSpecifier;
+
+              const importPackageExportName: string =
+                (exportSpecifier.propertyName || exportSpecifier.name).getText();
+
+              return {
+                symbol: current,
+                importPackagePath: packagePath,
+                importPackageExportName: importPackageExportName.trim(),
+                isAmbient: false
+              };
+            }
+          }
+        }
```

Comment:
>} [](start = 8, length = 1)

can you extract this into a function?

Reply:
Good idea

---
In reply to: [161947131](https://github.com/Microsoft/web-build-tools/pull/491#discussion_r161947131) [](ancestors = 161947131)